### PR TITLE
Implement AwaStaticClient_RegisterResourceWithPointer

### PR DESCRIPTION
**Pull request by chris-dewbery-imgtec**
This pull request adds support for AwaStaticClient_RegisterObject() and AwaStaticClient_RegisterResourceWithPointer(). 

I do however have some questions about the API in this regard.

### CODE REVIEW

File: core/src/client/lwm2m_static.c [hunk 1]
```diff
@@ -401,6 +488,11 @@ AwaError AwaStaticClient_RegisterResourceWithHandler(AwaStaticClient * client, c
             ResourceDefinition * resourceDefinition = Definition_NewResourceTypeWithHandler(objFormat, resourceName, resourceID, resourceType, minimumInstances, maximumInstances, operations, (LWM2MHandler)handler);
             if (resourceDefinition != NULL)
             {
+                // TODO: perhaps a better option would be to bind some opaque data
```
Comment (chris-dewbery-imgtec) [inline core/src/client/lwm2m_static.c]: The definition registry maintains a pointer to the data, and two integers to support each resource. I'm not sure this is the best approach. I was thinking perhaps I could allocate some memory in lwm2m_static to store this data when required and store a pointer to this in the definition registry instead, what do you guys think?

File: core/src/client/lwm2m_static.c [hunk 2]
```diff
@@ -318,6 +318,93 @@ int AwaStaticClient_Process(AwaStaticClient * client)
     return result;
 }
 
+static AwaLwm2mResult AwaStaticClientDefaultHandler(AwaStaticClient * client, AwaOperation operation, AwaObjectID objectID,
```
Comment (chris-dewbery-imgtec) [inline core/src/client/lwm2m_static.c]: Having a single handler for both object and resource operations seems a little clunky to me, is this the intended use of the API?

File: core/src/client/lwm2m_static.c [hunk 3]
```diff
@@ -318,6 +318,93 @@ int AwaStaticClient_Process(AwaStaticClient * client)
...
+            case AwaOperation_DeleteResource:
+                break;
+
+            case AwaOperation_Write:
+            
+                resourceDefinition = Definition_LookupResourceDefinitionFromObjectDefinition(objectDefinition, resourceID);
+                if (resourceDefinition != NULL)
+                {
+                // TODO: check dataSize vs definition->DataElementSize
+                // TODO: what do we do about storing the length.. for opaque etc...
```
Comment (chris-dewbery-imgtec) [inline core/src/client/lwm2m_static.c]: The "WithPointer" functions make the API nice and clean for primative types, i.e integers, floats, but where are we supposed to store the length for opaque.

When we register a resource we supply a single pointer to the data. which in most cases is ok, however for the opaque case we need to store data + length.

Now we could use a pointer to a AwaOpaque type, which contains data and length. but in this case data is a void pointer, where we really need something statically allocated.

File: core/src/client/lwm2m_static.c [hunk 4]
```diff
@@ -318,6 +318,93 @@ int AwaStaticClient_Process(AwaStaticClient * client)
...
+                    result = AwaLwm2mResult_BadRequest;
+                }
+                break;
+
+            case AwaOperation_Read:
+
+                resourceDefinition = Definition_LookupResourceDefinitionFromObjectDefinition(objectDefinition, resourceID);
+                if (resourceDefinition != NULL)
+                {
+                    offset = resourceDefinition->DataPointer + (resourceDefinition->DataStepSize * objectInstanceID);
```
Comment (chris-dewbery-imgtec) [inline core/src/client/lwm2m_static.c]: There are 3 parameters in AwaStaticClient_RegisterResourceWithPointer() the dataPointer, dataElementSize and dataStepSize. The code above assumes that the step size will be set to the element size for sequential data (i.e array types) and will be set to the sizeof the structure when using structures. 

for example: 

`typedef struct
{
    char Manufacturer[64];
    float Temperature;

} HeaterObject;

static HeaterObject heater[HEATER_INSTANCES];

AwaStaticClient_RegisterObject(awaClient, "Heater", 1000, 0, HEATER_INSTANCES);
AwaStaticClient_RegisterResourceWithPointer(awaClient, "Manufacturer", 1000, 101, AwaResourceType_String, 0, 1, AwaAccess_Read,
                                                &heater[0].Manufacturer, sizeof(heater[0].Manufacturer), sizeof(heater[0]));
AwaStaticClient_RegisterResourceWithPointer(awaClient, "Temperature",  1000, 104, AwaResourceType_Float, 0, 1, AwaAccess_Read,
                                                &heater[0].Temperature, sizeof(heater[0].Temperature), sizeof(heater[0]));

`

File: core/src/client/lwm2m_static.c (hunk 3 above)
Comment (seank-img) [inline core/src/client/lwm2m_static.c]: We have to store how much of the opaque buffer is used, I don't think there is a way around that. The user of the API is supplying us with the maximum amount of data (perhaps we need to rename dataElementSize to dataMaximumElementSize) that can be stored in each element. We might need a function to set/get how much of the opaque data element has been used for a given resource instance.

File: core/src/client/lwm2m_static.c (hunk 2 above)
Comment (seank-img) [inline core/src/client/lwm2m_static.c]: Agreed - for now we can just have two typedefs and refine it later?

**Pull request merged**